### PR TITLE
DS-606 dependency setup

### PR DESCRIPTION
**Pull request by augustinas**
The PR is aimed at supporting the transfer of dependency setup logic from frontend components to the backend application service. Change include:

- addition of dependency setup logic
- exceptions specific to the application setup failure mode
- fixes to the previously failing tests

Notable service behaviour: overall application setup process does not fail if one of the dependencies fail to setup. This might change in the future.

### CODE REVIEW

File: hat/app/org/hatdex/hat/she/mappers/DataEndpointMapper.scala [hunk 1]
```diff
@@ -110,10 +110,10 @@ trait DataEndpointMapper extends JodaWrites with JodaReads {
           s"- ${t.minusMinutes(1).toString("dd MMMM")}"
         }
         else {
-          s"- ${t.withZone(start.getZone).toString("dd MMMM HH:mm z")}"
+          s"- ${t.withZone(start.getZone).toString("dd MMMM HH:mm ZZZ")}"
```
Comment (whiteshadow-gr) [inline hat/app/org/hatdex/hat/she/mappers/DataEndpointMapper.scala:113]: Why the change in format? I wonder how the front end will react to that

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala [hunk 2]
```diff
@@ -125,42 +128,91 @@ class ApplicationsService @Inject() (
...
       logger.debug(s"Application status for ${app.application.id}: active = ${app.active}, setup = ${app.setup}, needs updating = ${app.needsUpdating}")
       app
     }
+  }
+
+  private def setupDependencies(application: HatApplication)(implicit hat: HatServer, user: HatUser, requestHeader: RequestHeader): Future[Option[Boolean]] = {
+    val maybeDependenciesSetup: Option[Future[Boolean]] = application.application.dependencies.map { deps =>
+      implicit val hatDb = hat.db
+
+      val appDeps = deps.plugs.toSet ++ deps.plugs.toSet
```
Comment (whiteshadow-gr) [inline hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:167]: Why ++ the same thing? This duplicates the records no?

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala [hunk 3]
```diff
@@ -125,42 +128,91 @@ class ApplicationsService @Inject() (
...
+      val toolDeps = deps.tools.toSet
+
+      val dependenciesEnabled = for {
+        appStatuses <- Future.sequence(appDeps.map(applicationStatus(_)))
+        updatedAppStatuses <- Future.sequence(appStatuses.map(_.get).filterNot(app => app.setup && app.enabled).map(setupApplication(_)))
```
Comment (whiteshadow-gr) [inline hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:172]: `.get` you cut some corners here didn't you?

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala [hunk 4]
```diff
@@ -125,42 +128,91 @@ class ApplicationsService @Inject() (
...
+        toolStatuses <- Future.sequence(toolDeps.map(fId => functionService.get(fId)))
+        updatedToolStatuses <- Future.sequence(toolStatuses.map(_.get).filterNot(_.status.enabled).map(f => functionService.save(f.copy(status = f.status.copy(enabled = true)))))
```
Comment (whiteshadow-gr) [inline hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:174]: `.get` here as well

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala [hunk 5]
```diff
@@ -125,42 +128,91 @@ class ApplicationsService @Inject() (
...
+          throw HatApplicationDependencyException(application.application.id, e.getMessage)
+
+        case _: NoSuchElementException =>
+          logger.warn(s"Dependency not found. Mis-configured application ${application.application.id}")
+          throw HatApplicationDependencyException(application.application.id, "Dependency not found")
+      }
+    }
+
+    maybeDependenciesSetup match {
+      case Some(value) => value.map(Some(_))
```
Comment (whiteshadow-gr) [inline hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:189]: minor thing. value can be renamed to `depSetup` or `isDepSetup`

File: hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala [hunk 6]
```diff
@@ -261,7 +262,39 @@ class ApplicationsServiceSpec(implicit ee: ExecutionEnv) extends PlaySpecificati
...
+      result must throwA[HatApplicationSetupException].await(1, 20.seconds)
+    }
+  }
+
+  "Application `setup` method for applications with dependencies" should {
+    "Enable plug dependencies" in {
+      val service = application.injector.instanceOf[ApplicationsService]
+      val result = for {
+        app <- service.applicationStatus(notablesAppDebitlessWithPlugDependency.id)
+        setup <- service.setup(app.get)
```
Comment (whiteshadow-gr) [inline hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala:274]: `get`

File: hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala [hunk 7]
```diff
@@ -261,7 +262,39 @@ class ApplicationsServiceSpec(implicit ee: ExecutionEnv) extends PlaySpecificati
...
+      }
+
+      result await (1, 20.seconds)
+    }
+
+    "Return partial success for application with invalid dependencies" in {
+      val service = application.injector.instanceOf[ApplicationsService]
+      val result = for {
+        app <- service.applicationStatus(notablesAppDebitlessWithInvalidDependency.id)
+        setup <- service.setup(app.get)
```
Comment (whiteshadow-gr) [inline hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala:290]: `get`

File: hat/test/org/hatdex/hat/she/functions/DataFeedDirectMapperSpec.scala [hunk 8]
```diff
@@ -77,7 +77,7 @@ class DataFeedDirectMapperSpec extends PlaySpecification with Mockito with DataF
       transformed.source must be equalTo "google"
       transformed.types must contain("event")
       transformed.title.get.text must contain("MadHATTERs Tea Party: The Boston Party")
-      transformed.title.get.subtitle.get must contain("12 December 18:30 - 22:30 EST")
+      transformed.title.get.subtitle.get must contain("12 December 18:30 - 22:30 America/New_York")
```
Comment (whiteshadow-gr) [inline hat/test/org/hatdex/hat/she/functions/DataFeedDirectMapperSpec.scala:80]: Just a comment. Really don't like our tests that involve time...

File: hat/app/org/hatdex/hat/she/mappers/DataEndpointMapper.scala (hunk 1 above)
Comment (augustinas) [reply hat/app/org/hatdex/hat/she/mappers/DataEndpointMapper.scala:113]: This was due to undocumented behaviour in the Joda library causing the tests to fail. I was unable to find a more satisfactory solution that would have the same output.

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala (hunk 2 above)
Comment (augustinas) [reply hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:167]: Good catch!

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala (hunk 3 above)
Comment (augustinas) [reply hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:172]: This is done purposefully - I wanted the error to be thrown in case any of the applications are not found. In previous commits `flatten` was used but that resulted in incorrect behaviour - failures were silently ignored.

File: hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala (hunk 4 above)
Comment (augustinas) [reply hat/app/org/hatdex/hat/api/service/applications/ApplicationsService.scala:174]: Same as above

File: hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala (hunk 6 above)
Comment (augustinas) [reply hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala:274]: I deem it acceptable in the tests. The only way it will throw None.get is if the test is setup wrongly and I see now way of gracefully recovering the wrong test :)

File: hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala (hunk 7 above)
Comment (augustinas) [reply hat/test/org/hatdex/hat/api/service/applications/ApplicationsServiceSpec.scala:290]: Same as above

File: hat/test/org/hatdex/hat/she/functions/DataFeedDirectMapperSpec.scala (hunk 8 above)
Comment (augustinas) [reply hat/test/org/hatdex/hat/she/functions/DataFeedDirectMapperSpec.scala:80]: Agreed
